### PR TITLE
MINOR: Remove `toStruct` and `fromStruct` methods from generated protocol classes

### PR DESCRIPTION
**Pull request by ijuma**
Update few classes that were still using the removed methods (including
tests that are no longer required).

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java [hunk 1]
```diff
@@ -188,9 +185,7 @@ public KafkaPrincipal deserialize(byte[] bytes) {
             throw new SerializationException("Invalid principal data version " + version);
```
Comment (chia7712) [inline clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java:185]: Could we use ```HIGHEST_SUPPORTED_VERSION``` and ```LOWEST_SUPPORTED_VERSION``` to replace origin condition ("0" and ```DefaultPrincipalData.SCHEMAS.length```) ?

**Review by chia7712 — APPROVED** [review]
@ijuma Nice cleanup!

File: clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java (hunk 1 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java:185]: Good point, addressed.

**Review by chia7712 — APPROVED** [review]

File: clients/src/test/java/org/apache/kafka/common/requests/OffsetFetchResponseTest.java [hunk 2]
```diff
@@ -107,20 +107,21 @@ public void testStructBuild() {
...
         for (short version = 0; version <= ApiKeys.OFFSET_FETCH.latestVersion(); version++) {
-            Struct struct = latestResponse.data().toStruct(version);
+            OffsetFetchResponseData data = new OffsetFetchResponseData(
+                new ByteBufferAccessor(latestResponse.serialize(version)), version);
 
-            OffsetFetchResponse oldResponse = OffsetFetchResponse.parse(latestResponse.serialize(version), version);
+            OffsetFetchResponse oldResponse = new OffsetFetchResponse(data, version);
 
             if (version <= 1) {
-                assertFalse(struct.hasField(ERROR_CODE));
```
Comment (chia7712) [inline clients/src/test/java/org/apache/kafka/common/requests/OffsetFetchResponseTest.java:115]: the field "ERROR_CODE" is never used.

File: clients/src/main/resources/common/message/README.md [hunk 3]
```diff
@@ -157,9 +157,6 @@ meaning of the request, such as a "validateOnly" boolean, should not be ignored.
 It's often useful to know how much space a message will take up before writing
 it out to a buffer.  You can find this out by calling the Message#size method.
 
-You can also convert a message to a Struct by calling Message#toStruct.  This
-allows you to use the functions that serialize Structs to buffers.
-
 Deserializing Messages
 ----------------------
 Message objects may be deserialized using the Message#read method.  This method
```
Comment (chia7712) [inline clients/src/main/resources/common/message/README.md:162]: There is another invalid description.

```
You can also deserialize a message from a Struct by calling Message#fromStruct.
The Struct will not be modified.
```

**Comment by dengziming** [comment]
This is nice, maybe we will also remove `struct.writeTo` and `struct.read` after #9641  #9766  and KAFKA-10863 is accomplished.

**Comment by ijuma** [comment]
@dengziming Looks like `StickyAssignor` also uses `Struct` still. Once all usages are removed, maybe we can delete `Struct` altogether instead of only some methods in it.

**Comment by ijuma** [comment]
Hmm, we may need to keep `Struct` as a type for `Schema`, however. So you may be right that we can only remove the serialization/deserialization methods.

**Comment by dengziming** [comment]
Yes, #9766 is trying to replace Struct in`StickyAssignor`. I am not sure about removing Struct class, I will try to check it, but we can remove serialization/deserialization methods.

**Comment by ijuma** [comment]
JDK 11 build passed, JDK 8 and 15 had unrelated flaky failures.

**Pull request merged**